File: MG/App/Backend/mojGradApp/mojGradApp/Models/InfoModels/PostInfo.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace mojGradApp.Models.InfoModels
{
    public class PostInfo
    {
        public int Id { get; set; }
        public int UserID { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public int CategoryID { get; set; }
        public DateTime Time { get; set; }
        public string TimePassed { get; set; }
        public string Location { get; set; }
        public int TypeID { get; set; }
        public string ImagePath { get; set; }
        public int CityID { get; set; }
        public double Longitude { get; set; }
        public double Latitude { get; set; }

        public string Username { get; set; }
        public int ReactionsNumber { get; set; }
        public string TypeName { get; set; }
        public int CommentsNumber { get; set; }
        public string CategoryName { get; set; }
        public string UserProfilePhoto { get; set; }
        public int ReportCount { get; set; }
        public int IsReacted { get; set; }
        public int IsReported { get; set; }
        public int IsSaved { get; set; }

        public PostInfo(Posts post)
        {
            this.Id = post.Id;
            this.UserID = post.UserID;
            this.Title = post.Title;
            this.Description = post.Description;
            this.CategoryID = post.CategoryID;
            this.Time = post.Time;
            this.TimePassed = GetTimePassed(this.Time);
            this.Location = post.Location;
            this.TypeID = post.TypeID;
            this.ImagePath = post.ImagePath;
            this.CityID = post.CityID;
            this.Username = post.User.Username;
            this.ReactionsNumber = post.PostReactions.Count();
            this.TypeName = post.Type.Name;
            this.CommentsNumber = post.Comments.Count();
            this.CategoryName = post.Category.Name;
            this.UserProfilePhoto = post.User.ProfilePhotoPath;
            this.Longitude = post.Longitude;
            this.Latitude = post.Latitude;
            this.ReportCount = post.PostReports.Count(x => x.PostID == this.Id);
        }

        public PostInfo(Posts post, int userID) : this(post)
        {
            if (post.PostReactions.FirstOrDefault(x => x.PostID == post.Id && x.UserID == userID) != null)
                this.IsReacted = 1;
            else
                this.IsReacted = 0;

            if (post.PostReports.FirstOrDefault(x => x.PostID == post.Id && x.UserID == userID) != null)
                this.IsReported = 1;
            else
                this.IsReported = 0;

            /*if (post.SavedPost.FirstOrDefault(x => x.PostID == post.Id && x.UserID == userID) != null)
                this.IsSaved = 1;
            else
                this.IsSaved = 0;*/
        }

        private string GetTimePassed(DateTime dt)
        {
            string timePassed;
            TimeSpan ts = DateTime.Now.Subtract(dt);
            timePassed = ts.Seconds + "s";
            if (ts.TotalSeconds > 60)
            {
                timePassed = ts.Minutes + "min";
                if (ts.TotalMinutes > 60)
                {
                    timePassed = ts.Hours + "h";
                    if (ts.TotalHours > 24)
                    {
                        timePassed = ts.Days + "d";
                        if (ts.TotalDays > 31)
                        {
                            timePassed = (int)(ts.TotalDays / 30) + "m";
                            if (ts.TotalDays > 365)
                            {
                                timePassed = (int)(ts.TotalDays / 365) + "y";
                            }
                        }
                    }
                }
            }
            return timePassed;
        }
    }
}

File: MG/App/Frontend/moj_grad_app/lib/models/modelsViews/postInfo.dart
class PostInfo {
  int id;
  int userId;
  String title;
  String description;
  int categoryId;
  DateTime time;
  String location;
  int typeId;
  String imagePath;
  double longitude;
  double latitude;
  int cityID;

  String username;
  int reactionsNumber;
  String typeName;
  int commentsNumber;
  String categoryName;
  String userProfilePhoto;
  int reportCount;
  int isReacted;
  int isReported;
  String timePassed;

  bool get getIsReacted {
    if (isReacted == 1)
      return true;
    else
      return false;
  }

  bool get getIsReported {
    if (isReported == 1)
      return true;
    else
      return false;
  }

  PostInfo(
      userId,
      title,
      description,
      categoryId,
      time,
      location,
      typeId,
      username,
      reactionsNumber,
      typeName,
      commentsNumber,
      categoryName,
      imagePath,
      userProfilePhoto,
      longitude,
      latitude,
      cityID,
      reportCount,
      isReacted,
      isReported,
      timePassed) {
    this.userId = userId;
    this.title = title;
    this.description = description;
    this.categoryId = categoryId;
    this.time = time;
    this.location = location;
    this.typeId = typeId;
    this.username = username;
    this.reactionsNumber = reactionsNumber;
    this.typeName = typeName;
    this.commentsNumber = commentsNumber;
    this.categoryName = categoryName;
    this.imagePath = imagePath;
    this.userProfilePhoto = userProfilePhoto;
    this.longitude = longitude;
    this.latitude = latitude;
    this.cityID = cityID;
    this.reportCount = reportCount;
    this.isReacted = isReacted;
    this.isReported = isReported;
    this.timePassed = timePassed;
  }

  Map<String, dynamic> toMap() {
    var map = Map<String, dynamic>();
    if (id != null) map['id'] = id;
    map['userID'] = userId;
    map['title'] = title;
    map['description'] = description;
    map['categoryID'] = categoryId;
    map['time'] = time.toIso8601String();
    map['location'] = location;
    map['typeID'] = typeId;
    map['username'] = username;
    map['reactionsNumber'] = reactionsNumber;
    map['typeName'] = typeName;
    map['commentsNumber'] = commentsNumber;
    map['categoryName'] = categoryName;
    map['imagePath'] = imagePath;
    map['userProfilePhoto'] = userProfilePhoto;
    map['longitude'] = longitude;
    map['latitude'] = latitude;
    map['cityID'] = cityID;
    map['reportCount'] = reportCount;
    map['isReacted'] = isReacted;
    map['isReported'] = isReported;
    map['timePassed'] = timePassed;
    return map;
  }

  PostInfo.fromObject(dynamic o) {
    this.id = o['id'];
    this.userId = o['userID'];
    this.title = o['title'];
    this.description = o['description'];
    this.categoryId = o['categoryID'];
    //DateTime date = DateTime.tryParse(o['time']);
    this.time = DateTime.tryParse(o['time']);
    this.location = o['location'];
    this.typeId = o['typeID'];
    this.username = o['username'];
    this.reactionsNumber = o['reactionsNumber'];
    this.typeName = o['typeName'];
    this.commentsNumber = o['commentsNumber'];
    this.categoryName = o['categoryName'];
    this.imagePath = o['imagePath'];
    this.userProfilePhoto = o['userProfilePhoto'];
    this.longitude = o['longitude'];
    this.latitude = o['latitude'];
    this.cityID = o['cityID'];
    this.reportCount = o['reportCount'];
    this.isReacted = o['isReacted'];
    this.isReported = o['isReported'];
    this.timePassed = o['timePassed'];
  }
}
